notebooks/06a Modelling - classification.ipynb

```diff
@@ -0,0 +1,865 @@
+   "source": [
+    "model_name = \"xgboost\"\n",
+    "\n",
+    "# define gridsearch parameters\n",
+    "gsc = GridSearchCV(\n",
+    "    estimator=XGBClassifier(\n",
+    "        weight=weights.to_dict(), random_state=42, tree_method=\"gpu_hist\"\n",
+    "    ),\n",
+    "    param_grid={\n",
+    "        \"n_estimators\": [50, 100, 200],  # default 100?\n",
+    "        \"learning_rate\": [0.01, 0.1, 1],  # default 0.3\n",
+    "        \"max_depth\": [5, 10],  # default 6\n",
+    "    },\n",
+    "    cv=5,\n",
+    "    scoring=\"f1_weighted\",\n",
+    "    verbose=1,\n",
+    "    n_jobs=-1,\n",
+    "    refit=True,\n",
+    ")\n",
+    "\n",
+    "# takes ~25 mins to run on a STANDARD_D13_V2\n",
+    "# takes ~52mins to run on a STANDARD_NC6\n",
+    "models[model_name] = train_model(gsc, X_train, y_train)\n",
+    "models[model_name]"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Save models"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "gather": {
+     "logged": 1653053267006
+    }
+   },
+   "outputs": [],
+   "source": [
+    "# save models outside the git tree\n",
+    "with open(\"../../models/classification.pickle\", \"wb\") as handle:\n",
+    "    pickle.dump(models, handle)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "nteract": {
+     "transient": {
+      "deleting": false
+     }
+    }
+   },
+   "source": [
+    "## Load models"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "gather": {
+     "logged": 1653053278650
+    },
+    "jupyter": {
+     "outputs_hidden": false,
+     "source_hidden": false
+    },
+    "nteract": {
+     "transient": {
+      "deleting": false
+     }
+    }
+   },
+   "outputs": [],
+   "source": [
+    "# load models from outside the git tree\n",
+    "with open(\"../../models/classification.pickle\", \"rb\") as handle:\n",
+    "    models = pickle.load(handle)\n",
+    "models"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Validate models\n",
+    "\n",
+    "Use the held-out test set to evaluate the performance of all the tuned models"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "gather": {
+     "logged": 1653053306159
+    },
+    "jupyter": {
+     "outputs_hidden": false,
+     "source_hidden": false
+    },
+    "nteract": {
+     "transient": {
+      "deleting": false
+     }
+    }
+   },
+   "outputs": [],
+   "source": [
+    "for model in models:\n",
+    "    preds = models[model][\"model\"].predict(X_test)\n",
+    "    probs = models[model][\"model\"].predict_proba(X_test)\n",
+    "    # calculate performance\n",
+    "    balanced_accuracy = balanced_accuracy_score(y_test, preds)\n",
+    "    f1_score_weighted = f1_score(y_test, preds, average=\"weighted\")\n",
+    "    auc = roc_auc_score(\n",
```

Comment:
Could also be worth looking at how precision/recall/f1 score vary by class, to see if this classifier is particularly strong/weak in specific classes

Reply:
Am adding actual vs predicted plots to all models for now, with note under "Extensions" to add PR curves per class in notebook